### PR TITLE
Do not override final Controller method render()

### LINKED ISSUE #4701: ControllerTrait::render() is final since Symfony 3.4
Seen while upgrading a Symfony app to 3.4:

```
The "Symfony\Bundle\FrameworkBundle\Controller\ControllerTrait::render()" method
is considered final since version 3.4. It may change without further notice as of
its next major version. You should not extend it from
"Sonata\AdminBundle\Controller\CRUDController".
```

### PR DESCRIPTION
**Pull request by hanzi**
I am targeting this branch, because this fixes a deprecation warning in Symfony 3.4 and a possible incompatibility with Symfony 4.0+.

Closes #4701 

## Changelog

```markdown
### Added
- Added `CRUDController::renderWithExtraParams` as a replacement for the `render` method

### Deprecated
- Deprecated `CRUDController::render`

### Fixed
- Deprecation warning for overriding `Controller::render` which is supposed to be final
```

## Subject

The basic solution has already been merged previously, in #4765.
It has then been reverted again in #4780 because it introduced a BC issue due to the now different behaviour of `CRUDController::render` (which was the inherited method from `Controller` that does not add required parameters.)

This PR re-implements the solution while avoiding the BC issues:

**Symfony 3.4 / 4.0+:** Instead of inheriting from `Controller`, we use the `ControllerTrait` (which contains all the helper methods) directly, aliasing the `render` method. No problem here.

**Symfony 3.3**: Same, but in 3.3 the `ControllerTrait` does not contain the methods `get()` and `has()`. I solve that issue by implementing `__call()` on `CRUDController` which handles those two methods specifically.

**Symfony 2.8 - 3.2:** These versions do not yet contain a `ControllerTrait`, so I polyfill it. That polyfill's methods are copied from 3.2's `Controller` class.

### CODE REVIEW

File: src/Controller/CRUDController.php [hunk 1]
```diff
@@ -37,11 +39,21 @@
 use Symfony\Component\Security\Core\Exception\AccessDeniedException;
 use Symfony\Component\Security\Csrf\CsrfToken;
 
+// BC for Symfony < 3.3 where this trait does not exist
+// NEXT_MAJOR: Remove the polyfill and inherit from Controller again
+require_once __DIR__.'/PolyfillControllerTrait.php';
```
Comment (Soullivaneuh) [inline src/Controller/CRUDController.php]: I'm not a big fan of vendor code copy. There is no another way? :confused:

File: src/Controller/CRUDController.php [hunk 2]
```diff
@@ -61,10 +73,60 @@ public function setContainer(ContainerInterface $container = null)
         $this->configure();
     }
 
+    // BC for Symfony 3.3 where ControllerTrait exists but does not contain get() and has() methods.
+    public function __call($method, $arguments)
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]: You will have to replace `is_callable` with `method_exists` somewhere in the code to fix the tests, look for it

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (greg0ire) [reply src/Controller/CRUDController.php]: The other way I know of is dropping sf 2. This is ugly, but should work fairly well and is quite maintainable since we are not supposed to edit `PolyfillControllerTrait` at all

File: src/Controller/CRUDController.php [hunk 3]
```diff
@@ -61,10 +73,60 @@ public function setContainer(ContainerInterface $container = null)
...
+     *
+     * @return Response A Response instance
+     *
+     * @deprecated
      */
     public function render($view, array $parameters = [], Response $response = null)
+    {
+        @trigger_error('Method '.__CLASS__.'::render has been renamed to '.__CLASS__.'::renderWithExtraParams.'
+            .' If you need to render a template WITHOUT the additional parameters, call the templating component'
+            .' directly, such as: $this->container->get(\'twig\')->render(...);', E_USER_DEPRECATED
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]: There is no way to call `render` before your PR anyway, so I'm not sure this tip should be included.

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (hanzi) [reply src/Controller/CRUDController.php]: I think it might be possible using Reflection: https://3v4l.org/ZEjh5

Problem is, `CRUDController` implements `__call` itself to fix Symfony 3.3 compatibility issues. Argh!

File: src/Controller/PolyfillControllerTrait.php [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        if (!$this->container->has('security.csrf.token_manager')) {
+            throw new \LogicException('CSRF protection is not enabled in your application.');
+        }
+
+        return $this->container->get('security.csrf.token_manager')->isTokenValid(new CsrfToken($id, $token));
+    }
+}
+
+
+if (!trait_exists('\Symfony\Bundle\FrameworkBundle\Controller\ControllerTrait')) {
```
Comment (greg0ire) [inline src/Controller/PolyfillControllerTrait.php]: Please use the `::class` notation

File: src/Controller/CRUDController.php (hunk 2 above)
Comment (greg0ire) [reply src/Controller/CRUDController.php]: Uuuh wait nevermind, that was in case `__call` was used to replace `render` Travis passes.

Comment (hanzi) [reply src/Controller/CRUDController.php]: Yes, I "borrowed" some parts from your PR, including that one. See the new line 522:
https://github.com/sonata-project/SonataAdminBundle/pull/4785/files#diff-a8a33ea35d3afb5f4ce4a69f18516b03R522

Comment (greg0ire) [reply src/Controller/CRUDController.php]: Oh right didn't see... and it is needed, I was wrong

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (Soullivaneuh) [reply src/Controller/CRUDController.php]: @greg0ire You mean SF 3.2? I think we can. It's BC and SF 3.2 is not maintained except for security fixes which will BTW end at the end of the year.

Comment (greg0ire) [reply src/Controller/CRUDController.php]: No, I mean sf 2.8. In sf 2, `render` is `public`, and can't be "overriden" with `__call` apparently.

File: src/Controller/CRUDController.php [hunk 5]
```diff
@@ -37,11 +39,21 @@
...
+require_once __DIR__.'/PolyfillControllerTrait.php';
+
 /**
  * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
  */
-class CRUDController extends Controller
+class CRUDController implements ContainerAwareInterface
 {
+
+    // NEXT_MAJOR: Don't use these traits anymore (inherit from Controller instead)
```
Comment (OskarStark) [inline src/Controller/CRUDController.php]: please use a FQCN instead of "Controller", otherwise it is not clear **which** controller to extend

File: src/Controller/PolyfillControllerTrait.php [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+ * The methods in this trait are a copy of the Controller class
+ * in the FrameworkBundle of Symfony 3.2.
+ *
+ * NEXT_MAJOR: Remove this file.
+ *
+ * @author Christian Kraus <christian@kraus.work>
+ */
+trait PolyfillControllerTrait
+{
+
```
Comment (OskarStark) [inline src/Controller/PolyfillControllerTrait.php]: please remove blank line

File: src/Controller/PolyfillControllerTrait.php [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+        $response = new BinaryFileResponse($file);
+        $response->setContentDisposition(
+            $disposition,
+            $fileName === null ? $response->getFile()->getFilename() : $fileName
+        );
+
+        return $response;
+    }
+
+
```
Comment (OskarStark) [inline src/Controller/PolyfillControllerTrait.php]: please only one blank line after method

File: tests/Controller/CRUDControllerTest.php [hunk 8]
```diff
@@ -587,7 +587,7 @@ public function testRender()
         $this->parameters = [];
         $this->assertInstanceOf(
             'Symfony\Component\HttpFoundation\Response',
-            $this->controller->render('FooAdminBundle::foo.html.twig', [], null)
```
Comment (OskarStark) [inline tests/Controller/CRUDControllerTest.php]: shouldn't we keep them as legacy tests?

Comment (greg0ire) [reply tests/Controller/CRUDControllerTest.php]: Uuuh yeah. Please copy that from my failed `__call` PR (it's the `@expectedDeprecation` annotated part)

File: src/Controller/PolyfillControllerTrait.php (hunk 4 above)
Comment (hanzi) [reply src/Controller/PolyfillControllerTrait.php]: Cool. I did not know you could do that even for non-existing classes/traits/interfaces.

Comment (greg0ire) [reply src/Controller/PolyfillControllerTrait.php]: Yeah no autoload is triggered, it's really just for convenience

**Comment by hanzi** [comment]
So, after some issues with the unit tests I have successfully implemented a different solution that works without copying most of the methods in `Controller`.

I have also ~stolen~ adapted two unit tests, courtesy of @greg0ire  :)

File: tests/Controller/CRUDControllerTest.php [hunk 9]
```diff
@@ -3713,6 +3713,23 @@ public function testBatchActionWithRequesData()
...
+    {
+        $this->setExpectedException(
+            \LogicException::class,
+            'Call to undefined method Sonata\AdminBundle\Controller\CRUDController::doesNotExist'
+        );
+        $this->controller->doesNotExist();
+    }
+
+    /**
+     * @expectedDeprecation
```
Comment (greg0ire) [inline tests/Controller/CRUDControllerTest.php]: Please add the expected message

File: src/Controller/CRUDController.php [hunk 10]
```diff
@@ -967,6 +1022,18 @@ public function getRequest()
...
+     * Gets a container configuration parameter by its name.
+     *
+     * @param string $name The parameter name
+     *
+     * @return mixed
+     */
+    protected function getParameter($name)
+    {
+        return $this->container->getParameter($name);
+    }
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]: What is this new method? Something really needed?

File: src/Controller/PolyfillControllerTrait.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    public function render($view, array $parameters = [], Response $response = null)
+    {
+        return $this->__call('render', [$view, $parameters, $response]);
+    }
+
+    /**
+     * @internal
+     */
+    protected function proxyToControllerClass($methodName, $arguments)
```
Comment (greg0ire) [inline src/Controller/PolyfillControllerTrait.php]: should probably be final

File: src/Controller/CRUDController.php (hunk 10 above)
Comment (hanzi) [reply src/Controller/CRUDController.php]: That method is the only one *still* in the FrameworkBundle `Controller` class, even in Symfony 4.0.

Since all that polyfill stuff might be removed in the next major version, implementing this method explicitely is required to maintain BC.

**Comment by Soullivaneuh** [comment]
You workaround looks better, but it's still a quite durty workaround IMHO.

The next Symfony LTS (3.4) will be out at the end of this month and we can bump to it according to our policy.

Is it really worth it to have this trick? Can't we wait some days to bump to Symfony 3.4 instead?

**Comment by hanzi** [comment]
Symfony 3.4 (and 4.0) are the cause of this problem in the first place.

Also, why would you want to drop support for Symfony 2.8 as soon as 3.4 is released? Symfony 2.8 will be maintained another *year* (two for security fixes.) Two of my customers have LTS codebases and will probably need a considerable part of that transitional year to upgrade, so losing support for the AdminBundle prematurely would be somewhat disappoiting.

Anyway, it would not help: Symfony 2.8 and 3.2 both require the same workaround, since both do not have the `ControllerTrait`. So bumping to 3.2 wouldn't change a thing here.

**Comment by greg0ire** [comment]
My PR based on `__call` did work for sf > 3 , so that's not exactly right.
Your point about sf 2 sounds good though. I propose we wait to see if the maintenance cost is very hard, e.g. if this is the only hack we have to do or if it is really hard to decide whether to drop 2.

**Comment by Soullivaneuh** [comment]
The reason to drop SF 2.8 is part of our policy and was already debated. See here: https://github.com/sonata-project/SonataAdminBundle/blob/3.x/CONTRIBUTING.md#dependency-changes

In a nutshell, we have only constraint: Keep one Symfony LTS support.

We drop older version support if the BC tricks are too heavy, and it seems to be the case.

I suggest to wait end of November and bump to Symfony 3.2. It will be compliant with our policy.

After that, if @sonata-project/contributors think the BC trick is quite acceptable, we can do it this way.

**Comment by jordisala1991** [comment]
at least the last LTS version MUST be supported

**Comment by Soullivaneuh** [comment]
> at least the last LTS version MUST be supported

Yeah, this is why I suggest to wait some days before merging the PR (with version bump) and make a release.

**Comment by jordisala1991** [comment]
My opinion is to not drop SF 2.8 on this stable (and add 4.0), but clean up on master to prepare support for 3.4 and 4.0 only.

So if this works and does not break bc promise it is okay by me

**Comment by greg0ire** [comment]
I'd much rather drop sf2 on stable than not have sf4 on it. This would delay support for sf 4 a lot I think.

**Comment by Soullivaneuh** [comment]
> My opinion is to not drop SF 2.8 on this stable

This is not the debate and this was already debated on dev-kit issues.

Symfony is a dependency like the other, we can change the version constraint on stable. The only constraint is the "At least one LTS".

The question is: Is this BC trick worth it? Yes? Let's merge it and release. No? Let's wait a few day and bump to the minimum SF version to avoid the trait requirement trick (should be 3.2 if I understood well).

**Comment by greg0ire** [comment]
I think this should be closed in favor of #4772 or made similar, sorry @hanzi but 3 major versions at once is a bit much. Your clients will still be able to use old releases, they will not disappear.

**Comment by jordisala1991** [comment]
Can we make another twitter poll or similar @greg0ire before unleashing the rage on slack again?

**Comment by greg0ire** [comment]
A poll certainly won't hurt: https://twitter.com/greg0ire/status/933379564549615617

**Comment by hanzi** [comment]
> I think this should be closed in favor of #4772 or made similar

Note that #4772 will actually **not fix the problem**.

Your PR *appears* to work because the unit tests exclusively invoke `render()` externally. This happens to work well because `CRUDController` considers this method to be public whereas FrameworkBundle `Controller` defines it protected. So after renaming, external invocations will hit `__call`.

But in the real world, this method would usually be invoked from within a child class of `CRUDController`, and those would still hit the "original" `render()` method since they are allowed to access protected methods.

In short, that approach would lead to the same BC issues we had after the last attempt in #4765.

**Comment by greg0ire** [comment]
Hmm ok, didn't see that coming! :sweat_smile: Good to know.

**Comment by hanzi** [comment]
Yep. So in order to avoid ugly workarounds like this, the minimum version we'd have to require is Symfony 3.3. And that one still needs the `__call` inside `CRUDController` which isn't the prettiest thing, either.

Anyway, I've added an additional check so that the polyfill file will not be loaded at all unless it's really needed.

**Comment by greg0ire** [comment]
@Soullivaneuh @jordisala1991 the poll shows that a big portion of your users still use 2.8, and that might be because of the user bundle. I don't think dropping 2 would be a good move until all our bundles support 3. Shall we merge this then?

**Comment by Soullivaneuh** [comment]
@greg0ire Honestly, I don't use user-bundle, so I'm not on the best place to make a decision. Both are doable, I let @sonata-project/contributors decide about it.

**Comment by greg0ire** [comment]
Decision taken, for the best I hope. @hanzi , thanks a lot for your patience!

**Comment by greg0ire** [comment]
Uuuuh I didn't notice there were that many commits! Next time please squash at least some of them :P

**Comment by hanzi** [comment]
@greg0ire  Sorry, will keep it in mind for the next PR!

Though, what is the best practice here? My concern was that squashing anything while the PR is in review might make it harder to follow the changes made in the meantime. Is it safe to just squash for every commit I make (so that, in effect, the PR always consists of only one commit)?

I assumed you'd just 'squash merge' the PR in the end.

**Comment by greg0ire** [comment]
I'd prefer that you keep it to one commit when necessary, and add a new commit if you do something unrelated. IIRC github can show changes between 2 pushes, I'm not sure it works with squashed commits though.
I wouldn't say it's the best practice, but it is ours. Ocramius will tell you otherwise if you contribute to Doctrine, so keep that in mind.

File: src/Controller/PolyfillControllerTrait.php [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Polyfills the FrameworkBundle's original ControllerTrait.
+ *
+ * NEXT_MAJOR: Remove this file.
+ *
+ * @author Christian Kraus <hanzi@hanzi.cc>
+ */
+trait PolyfillControllerTrait
```
Comment (core23) [inline src/Controller/PolyfillControllerTrait.php]: This is a ugly fix, because it will ruin every autocompletion in most IDEs
